src/data/equiv/one_sided.lean

```diff
@@ -0,0 +1,215 @@
+by { cases f, refl, }
+
+/-- See Note [custom simps projection] -/
+def injection.simps.inv_fun (e : α ↬ β) : β → α := e.symm
+
+initialize_simps_projections injection (to_fun → apply, inv_fun → symm_apply)
+
+/-- See Note [custom simps projection] -/
+def surjection.simps.inv_fun (e : α ↠ β) : β → α := e.symm
+
+initialize_simps_projections surjection (to_fun → apply, inv_fun → symm_apply)
+
+@[simp] lemma injection.symm_apply_apply (e : α ↬ β) (x : α) : e.symm (e x) = x :=
+e.left_inv x
+
+@[simp] lemma surjection.apply_symm_apply (e : α ↠ β) (x : β) : e (e.symm x) = x :=
+e.right_inv x
+
+@[simp] lemma injection.symm_comp_self (e : α ↬ β) : e.symm ∘ e = id := funext e.symm_apply_apply
+
+@[simp] lemma surjection.self_comp_symm (e : α ↠ β) : e ∘ e.symm = id := funext e.apply_symm_apply
+
+end symm
+
+/-! ### Extensionality -/
+
+section ext
+
+@[ext] lemma injection.ext {f g : α ↬ β}
+  (H : ∀ x, f x = g x) (H_symm : ∀ x, f.symm x = g.symm x) : f = g :=
+begin
+  cases f,
+  cases g,
+  congr,
+  exact funext H,
+  exact funext H_symm,
+end
+
+@[ext] lemma surjection.ext {f g : α ↠ β}
+  (H : ∀ x, f x = g x) (H_symm : ∀ x, f.symm x = g.symm x) : f = g :=
+begin
+  cases f,
+  cases g,
+  congr,
+  exact funext H,
+  exact funext H_symm,
+end
+
+end ext
+
+/-! ### Composition and identity -/
+
+section comp
+
+namespace injection
+
+def id (α : Sort*) : α ↬ α := ⟨id, id, λ _, rfl⟩
+
+@[simp] lemma id_apply (x : α) : id α x = x := rfl
+
+def comp (f : β ↬ γ) (g : α ↬ β) : α ↬ γ:=
+⟨f ∘ g, g.inv_fun ∘ f.inv_fun, f.left_inv.comp g.left_inv⟩
+
+@[simp] lemma comp_id (f : α ↬ β) : f.comp (id α) = f := ext (λ _, rfl) (λ _, rfl)
+
+@[simp] lemma id_comp (f : α ↬ β) : (id β).comp f = f := ext (λ _, rfl) (λ _, rfl)
+
+@[simp] lemma comp_apply (f : β ↬ γ) (g : α ↬ β) (x : α) : (f.comp g) x = f (g x) := rfl
+
+@[simp] lemma symm_comp_apply (f : β ↬ γ) (g : α ↬ β) (x : γ) :
+  (f.comp g).symm x = g.symm (f.symm x) := rfl
+
+lemma comp_assoc (f : γ ↬ δ) (g : β ↬ γ) (h : α ↬ β) :
+  (f.comp g).comp h = f.comp (g.comp h) :=
+by ext; simp
+
+end injection
+
+namespace surjection
+
+def id (α : Sort*) : α ↠ α := ⟨id, id, λ _, rfl⟩
+
+@[simp] lemma id_apply (x : α) : id α x = x := rfl
+
+def comp (f : β ↠ γ) (g : α ↠ β) : α ↠ γ:=
+⟨f ∘ g, g.inv_fun ∘ f.inv_fun, f.right_inv.comp g.right_inv⟩
+
+@[simp] lemma comp_id (f : α ↠ β) : f.comp (id α) = f := ext (λ _, rfl) (λ _, rfl)
+
+@[simp] lemma id_comp (f : α ↠ β) : (id β).comp f = f := ext (λ _, rfl) (λ _, rfl)
+
+@[simp] lemma comp_apply (f : β ↠ γ) (g : α ↠ β) (x : α) : (f.comp g) x = f (g x) := rfl
+
+@[simp] lemma symm_comp_apply (f : β ↠ γ) (g : α ↠ β) (x : γ) :
+  (f.comp g).symm x = g.symm (f.symm x) := rfl
+
+lemma comp_assoc (f : γ ↠ δ) (g : β ↠ γ) (h : α ↠ β) :
+  (f.comp g).comp h = f.comp (g.comp h) :=
+by ext; simp
+
+end surjection
+
+end comp
+
+/-! ### Properties -/
+
+section aliases
+
+protected lemma injection.left_inverse (e : α ↬ β) : left_inverse e.symm e :=
```

Comment:
Do you want to also provide `has_left_inverse` and `has_right_inverse`?

Reply:
I don't really see any point - I figured I'd just expose the bits of API already exposed by `equiv`. Strictly the only necessary lemma here is this one, `injection.injective` is just a shorthand.